feat!(nuxt): upgrade nuxt module

pi0 opened the pull request:
Rework of #791

- Add pinna as a dependency -- Nuxt modules are supposed to be standalone and provide integration out of the box
- Add simple playground 
- Migrate to [module-builder](https://github.com/nuxt/module-builder) and [nuxt kit](https://v3.nuxtjs.org/guide/going-further/kit)
- Split vue 2 (bridge) and vue 3 plugins
- Remove deprecated `ctx.pinia` 
- Nuxt 2 support requires bridge to be enabled (using plain simple plugin)
- Use `nuxi typecheck` for tpyechecking module
- `vue-demi` is removed for vue2 support since bridge is in place
- Auto imported composables (`definePiniaStore()` and `usePinia()` exposed)

Breaking changes:
- Nuxt 2 users need to enable [nuxt bridge](https://github.com/nuxt/bridge) in order to use latest version of the module.
- Only `$pinia` injection is available which can be accessed using `usePinia()` (or `useNuxtApp().$pinia`)
- `$nuxt` from store can be accessed using `useNuxtApp()` composable

To be tested:
- [x] Nuxt 3 on an external project
- [x] Nuxt 2 (w, w/o bridge) on an external project

pi0 committed:
feat(nuxt)!: rewrite with nuxt/kit

pi0 committed:
auto imports

pi0 committed:
update deps

pi0 committed:
lint

pi0 committed:
use nuxi to typecheck nuxt

pi0 committed:
add vue-tsc devDependency

pi0 committed:
fix build script

pi0 committed:
remove vue-demi

pi0 committed:
simplify injections

pi0 committed:
feat: usePinia composable

pi0 committed:
refactor: use resolver

pi0 committed:
fix lint issues

posva commented on packages/nuxt/package.json:
```diff
@@ -24,35 +24,34 @@
   "dependencies": {
-    "vue-demi": "*"
+    "@nuxt/kit": "^3.0.0-rc.4",
+    "pinia": ">=2.0.14"
```
Is this really a good idea? Then, it will be more difficult to upgrade pinia for users. Given how package managers error when you don't install the peer dependencies it feels like having a peer dep is a better direction.

pi0 replied:
Nuxt itself has a direct dependency on `vue`, `vue-router` and other libs and Nuxt modules do the same. `<pm> upgrade @pinia/nuxt` updates all of its sub dependencies within supported semver range (here is 2.x) (therefore the latest version of pinia will be installed). This gives one package installation DX. It also helps in some rare cases that you directly control working dependency version without asking users to manually do.

Regardless I'm fine with either direction you like :)

posva commented on packages/nuxt/src/runtime/composables.ts:
```diff
@@ -0,0 +1,6 @@
+import { useNuxtApp } from '#imports'
+import { defineStore } from 'pinia'
+
+export const usePinia = () => useNuxtApp().$pinia
```
do we really need this? This sohuld be the same as `getActivePinia()`

pi0 replied:
Hmm then i guess not if we already have a composable to access it. Does it return same reference of `$pinia`? (then probably plugin injection is not necessary too). 

One small benefit of nuxt composables vs vue composables is that they also work outside of vue setup lifecycle with better accessibility support.

posva commented on packages/nuxt/src/module.ts:
```diff
@@ -1,70 +1,59 @@
+    const composables = resolver.resolve('./runtime/composables')
+    addAutoImport([
+      { from: composables, name: 'usePinia' },
+      { from: composables, name: 'definePiniaStore' },
```
Why the rename?

pi0 replied:
As of [module author guide](https://v3.nuxtjs.org/guide/going-further/modules#always-prefix-exposed-interfaces), having prefix avoids possible conflict between two modules exposing same name for auto imports.

posva replied:
hmm, I think in this case, pinia can keep the defineStore and others will use a prefix

pi0 replied:
[Global store RFC](https://github.com/nuxt/framework/discussions/571) is still open. If we add it as built-in and even if based on pinna for implementation, module and built-in might conflict and introduce slight breaking change. If you prefer to advocate `defineStore`, we can update example to add explicit `import { defineStore } from 'pinia'` :)

pi0 replied:
(make `definePiniaStore` opt-in since is not standard)

posva commented on packages/nuxt/src/module.ts:
```diff
@@ -1,70 +1,59 @@
-  // }
-}
-
-declare module '@nuxt/types' {
```
if we remove these we won't have access to `$nuxt.$pinia`. Unless nuxt is somehow adding the types magically ✨

pi0 replied:
Indeed it does infer when returning `{ provide }` from plugins :)

posva replied:
but what about Nuxt 2?

pi0 replied:
Having bridge enabled, same DX should be provided (ee need to use plugin wrapper for vue2 pligin in this case)

posva committed:
chore: add back the name and missing configs

posva committed:
wip: add incomplete missing types

pi0 commented on packages/nuxt/src/runtime/plugin.vue2.ts:
```diff
@@ -10,6 +10,13 @@ export default (context: any, inject: any) => {
+  // @ts-expect-error: _p is internal
+  pinia._p.push(({ store }) => {
+    // make it non enumerable so it avoids any serialization and devtools
+    Object.defineProperty(store, '$nuxt', { value: context })
```
Nuxt context (app) can be accessed with useNuxtApp() global. Any specific usecase for $nuxt as a plugin?

posva replied:
People use to access `$fetch` and other properties. Should they not? I was quite requested back when I added it

pi0 replied:
For Nuxt 2 yes made sense, but in Nuxt 3 (and with bridge), we have global `$fetch` and composables replacing the old `$nuxt` API.

posva replied:
Should I mark it as deprecated then? Is there a page that makes sense to point in the jsdoc?

pi0 replied:
It is (currently briefly) documented here: https://v3.nuxtjs.org/guide/going-further/nuxt-app

Since this PR is breaking change, I would drop it. `$nuxt` interface in Nuxt 2 is not same as `NuxtApp`. It might be confusing to keep (still your choice)

pi0 replied:
(to be deprecated)

pi0 commented on packages/nuxt/src/runtime/plugin.vue2.ts:
```diff
@@ -21,3 +28,13 @@ export default (context: any, inject: any) => {
+declare module 'pinia' {
+  export interface PiniaCustomProperties {
+    /**
+     * Nuxt context.
```
It is NuxtApp from @nuxt/schema if you need to manually extend...

posva commented:
I force pushed because due to conflicts in pnpm-lock

posva committed:
ci: prepare types

posva committed:
chore: format

posva committed:
test(ts): add src folder

posva committed:
chore: fix tsconfig

posva committed:
test: remove failing types

posva commented:
The types never worked on my end. It's as if vue-tsc didn't pick up the correct tsconfig so I removed them from CI

posva commented:
Do you also get this one:

![Screenshot 2022-07-08 at 17 38 27](https://user-images.githubusercontent.com/664177/178025022-84b72d68-dbc3-454c-8e8b-9560e3676593.png)

It didn't appear before...

pi0 commented:
Thanks for the review @posva. I will rebase and check last changes on bridge.

Aareksio commented on packages/nuxt/src/module.ts:
```diff
@@ -1,70 +1,62 @@
+    // Make sure we use the mjs build for pinia
+    nuxt.options.alias.pinia = 'pinia/dist/pinia.mjs'
+
+    // Add runtime plugin
+    if (isNuxt2()) {
+      addPlugin(resolver.resolve('./runtime/plugin.vue2'))
+    } else {
+      addPlugin(resolver.resolve('./runtime/plugin.vue3'))
+    }
```
> Note: By default plugin is prepended to the plugins array. You can use second argument to append (push) instead.

This means, that by default, if we precise an array of modules like so:
```
modules: ['@pinia/nuxt', 'my-module']
```

We'll receive `getActivePinia was called with no active Pinia` error, as any plugin added by `my-module` with default options is going to be executed **before** pinia plugin can call `setActivePinia()`.

See reproduction: https://stackblitz.com/edit/github-ba6rwp?file=nuxt.config.ts

---

Raised as issue in nuxt: https://github.com/nuxt/framework/issues/5835

pi0 replied:
You shouldn't depend on module order (please see my comment: https://github.com/nuxt/framework/issues/5835#issuecomment-1181679108). We might improve with a new hook to allow extending pinia via other module plugins.

pi0 committed:
add missing peer dep

pi0 commented on packages/nuxt/src/runtime/plugin.vue3.ts:
```diff
@@ -0,0 +1,37 @@
+  setActivePinia(pinia)
+
+  // @ts-expect-error: _p is internal
+  pinia._p.push(({ store }) => {
```
(TODO: Remove from nuxt 3)

posva committed:
feat(nuxt): remove wrong `$nuxt` in Nuxt 3

BREAKING CHANGE: in Nuxt 3, `$nuxt` is no longer available in stores.
This is because it was removed in Nuxt 3 and it is no longer the
_context_ as it used to be. Most of the features used there, like
`$fetch` are now globally available and therefore remove the need of it.
You can also use
[`useNuxtApp()`](https://v3.nuxtjs.org/bridge/bridge-composition-api/)
when necessary.

posva committed:
feat(nuxt): deprecate old `$nuxt` context

BREAKING CHANGE: Starting on this version, `@pinia/nuxt` only works with
Nuxt 2 + Bridge and Nuxt 3, it no longer works with Nuxt 2 only. This is
necessary to have one single plugin that works well with the different
versions of Nuxt. If you aren't using bridge with Nuxt 2, check out the
[migration guide](https://v3.nuxtjs.org/bridge/overview) or pin your
`@pinia/nuxt` dependency in your:

```diff
-    "@pinia/nuxt": "^0.2.1",
+    "@pinia/nuxt": "0.2.1",
```

The `$nuxt` context usage should be replaced with globals like
`$fetch()` and `useNuxtApp()`. You can find more information about this
in Nuxt documentation:

- https://v3.nuxtjs.org/bridge/bridge-composition-api/
- https://v3.nuxtjs.org/bridge/overview

posva committed:
style: format comment

posva committed:
docs: remove outdated nuxt section

posva committed:
docs: updated nuxt instructions

posva committed:
feat(nuxt): add `autoImports` option in module

posva committed:
fix(nuxt): correct type for `$nuxt`

BREAKING CHANGE: `$nuxt` usage in stores defaults to type `any` unless
you install the `@nuxt/types` package. This is because that package is
quite heavy and can cause conflicts in projects not requiring it. Note
`$nuxt` is deprecated and shouldn't be used (cf the other breaking
changes notes).

posva committed:
docs: update version

posva committed:
chore: ok

posva committed:
chore: up lock
